src/librustc_error_codes/error_codes/E0308.md

```diff
@@ -13,7 +13,7 @@ let x: i32 = "I am not a number!";
 ```
 
 This error occurs when the compiler was unable to infer the concrete type of a
-variable. It can occur for several cases, the most common of which is a
-mismatch in the expected type that the compiler inferred for a variable's
-initializing expression, and the actual type explicitly assigned to the
-variable.
+variable. It can happen in several cases, the most common being a mismatch
+between the type that the compiler inferred for a variable based on its
+initializing expression, on the one hand, and the type the author explicitly
```

Comment:
Is this a typo? `on the one hand`

Reply:
Just my two cents, `on one hand` and `on the other hand` may not be easily understood by non-native english speakers. Maybe it's just me but I learned that late, `and` is easier to be understood.

Reply:
I meant to contribute a better message with no "on the other hand." But I got side-tracked and honestly a bit discouraged, since  there's nobody to workshop good messages with, or standards.

I'd like there to be a working group to bounce ideas off of, and/or a style guide.

Here's what I had, roughly:

```
This error occurs when the compiler was unable to infer the concrete type of a variable. It can
occur for several cases; The most common case is a mismatch between two types: the type explicitly assigned to a variable by the [code's] author, and the type the compiler infers for an expression assigned to that variable.
```

Possibly a "hint" line about dynamic typing versus static typing:

```
In Rust each declared variable is assigned a static type. Dynamic typing is not supported.
```

See prior discussion here if you want: https://users.rust-lang.org/t/looking-for-feedback-on-an-improved-error-message-for-e0308/40004

I actually think the phrasing can be much simplified.

Any two mis-matching tyoes assigned to the same variable cause this error. Regardless of explicit or inferred, at or after initialization, etc... Just... two non-matching types, one variable.

(Apparently there is a thing called [shadowing](https://doc.rust-lang.org/book/ch03-01-variables-and-mutability.html?highlight=shadow#shadowing), but if I read correctly, I believe it just spins up a new variable with a different compiled identity, but re-using the same name in source code. I am afraid of over-simplifying, and ending up with a message that is inaccurate regarding shadowing.)

Reply:
> This error occurs when the compiler was unable to infer the concrete type of a variable.
It can occur for several cases; The most common case is a mismatch between
two types: the type explicitly assigned to a variable by the [code's] author,
and the type the compiler infers for an expression assigned to that variable.

That is better than the merged one I think but still kinda long-winded, I think documentation should be short, concise and understandable. What I would say a good example is IETF RFC docs, rust docs is good enough for me.

Reply:
Thanks for the reply.

I have some things to do at the moment, but I will post that version as a second PR when I get the chance. Hopefully before a new stable release.

It could probably be shortened, mostly by removing the "common example" of when this error occurs. (It is kind of arbitrary, and I think it might be overly specific when the error is more general than that.) But I don't want to be too drastic while I am still getting used to contributing to Rust.

---

Posting this for reference since you mentioned IETF RFCs as being a good example: https://tools.ietf.org/html/rfc7322 The IETF RFC Style Guide.

Maybe that can be an influence for a Rust docs "style guide" or best practices guide. No time now, but planning to look that over later.